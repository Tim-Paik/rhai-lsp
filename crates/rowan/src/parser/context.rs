//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
