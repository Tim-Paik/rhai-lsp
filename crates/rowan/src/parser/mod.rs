use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context)
    }
}

/// The result of parsing.
#[derive(Debug, Clone)]
pub struct Parse {
    pub errors: Vec<ParseError>,
    pub green: GreenNode,
}

impl Parse {
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

#[derive(Debug, Clone, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    pub range: rowan::TextRange,
    pub kind: ParseErrorKind,
}

impl ParseError {
    pub fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

#[derive(Debug, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .intersperse(String::from(", "))
            .collect::<String>()
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
