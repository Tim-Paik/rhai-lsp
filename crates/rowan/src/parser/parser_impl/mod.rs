#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
