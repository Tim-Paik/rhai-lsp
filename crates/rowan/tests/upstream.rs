use rhai_rowan::parser::Parser;
use test_case::test_case;

#[test_case(include_str!("../../../rhai/scripts/array.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/assignment.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/comments.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/fibonacci.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl4.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/loop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/mat_mul.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/module.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/oop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/primes.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/speed_test.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/string.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/strings_map.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/switch.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/while.rhai"))]
fn parse_rhai_upstream(src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty());
    // TODO(tests)
    // insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
}
