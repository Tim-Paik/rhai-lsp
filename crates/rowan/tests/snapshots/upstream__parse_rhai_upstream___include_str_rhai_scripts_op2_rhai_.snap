---
source: crates/rowan/tests/upstream.rs
expression: "format!(\"{:#?}\", parse.into_syntax())"

---
FILE@0..69
  STMT@0..35
    ITEM@0..34
      EXPR@0..34
        EXPR_CALL@0..34
          EXPR@0..5
            EXPR_IDENT@0..5
              IDENT@0..5 "print"
          ARG_LIST@5..34
            PUNCT_PAREN_START@5..6 "("
            EXPR@6..33
              EXPR_LIT@6..33
                LIT@6..33
                  LIT_STR@6..33 "\"The result should be ..."
            PUNCT_PAREN_END@33..34 ")"
    PUNCT_SEMI@34..35 ";"
  WHITESPACE@35..37 "\n\n"
  STMT@37..57
    ITEM@37..56
      EXPR@37..56
        EXPR_LET@37..56
          KW_LET@37..40 "let"
          WHITESPACE@40..41 " "
          IDENT@41..42 "x"
          WHITESPACE@42..43 " "
          OP_ASSIGN@43..44 "="
          EXPR@44..56
            EXPR_BINARY@44..56
              WHITESPACE@44..45 " "
              EXPR_LIT@45..47
                LIT@45..47
                  LIT_INT@45..47 "12"
              WHITESPACE@47..48 " "
              OP_ADD@48..49 "+"
              EXPR@49..56
                EXPR_BINARY@49..56
                  WHITESPACE@49..50 " "
                  EXPR_LIT@50..52
                    LIT@50..52
                      LIT_INT@50..52 "34"
                  WHITESPACE@52..53 " "
                  OP_MUL@53..54 "*"
                  EXPR@54..56
                    WHITESPACE@54..55 " "
                    EXPR_LIT@55..56
                      LIT@55..56
                        LIT_INT@55..56 "5"
    PUNCT_SEMI@56..57 ";"
  WHITESPACE@57..59 "\n\n"
  STMT@59..68
    ITEM@59..67
      EXPR@59..67
        EXPR_CALL@59..67
          EXPR@59..64
            EXPR_IDENT@59..64
              IDENT@59..64 "print"
          ARG_LIST@64..67
            PUNCT_PAREN_START@64..65 "("
            EXPR@65..66
              EXPR_IDENT@65..66
                IDENT@65..66 "x"
            PUNCT_PAREN_END@66..67 ")"
    PUNCT_SEMI@67..68 ";"
  WHITESPACE@68..69 "\n"

